### DISCUSSION TITLE
Environment level parameters

### DISCUSSION THREAD

**Discussion by tegryan-ggs** [aws/copilot-cli#4181] [Ideas]
We leverage Copilot to deploy a lot of infrastructure, some of which accepts parameters which are different per environment.  An example might be SNS topics where you want to subscribe an email address to it, but it could be a different email in your production environments versus your test environments.  I know there is the addons.parameters.yml file to pass parameters to my addons stacks, but AFIK that only accepts static values or references to CF resources.

By adding a bit of bash to the buildspec.yml file I think we can overload that file with env specific parameters.

Replace:
```
for svc in $svcs; do
```

with:
```
          for svc in $svcs; do
          params_file=$CODEBUILD_SRC_DIR/copilot/$svc/addons/addons.parameters.yml
          env_file=$CODEBUILD_SRC_DIR/copilot/environments/$env/parameters.yml
          if [ -f "$params_file" ]; then
            if [ -f "$env_file" ]; then
              cat $env_file > $params_file
            fi
          fi
```

The drawback of this approach is that it only works in AWS Codebuild/Codepipeline deployments, and not with a `copilot svc deploy`.  Is there a way to move this functionality into Copilot proper so it works with both deployments?

**Comment by huanjani** [comment]
Hi, @tegryan-ggs! 

You should be able to achieve that by using [`Mappings`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/mappings-section-structure.html) to indicate changes in parameters among environments. See example here: https://github.com/aws/copilot-cli/issues/2432#issuecomment-857744698

Hope that helps!

Related: #3209, #2432.

**Comment by tegryan-ggs** [comment]
Hi Janice, I think there are two differences between the above and mappings.  (I might be wrong though).

1.  The above injects those parameters into every service in the environment
2.  The above keeps parameters defined in yml files.

I'll take a closer look at mappings, thanks!

**Comment by huanjani** [comment]
Ah, okay. So for environment-wide parameters via addons, the Copilot team is currently hard at work on env addons! In much the same way that you can add a `addons.parameters.yml` file to workload addons, you will soon be able to add a parameters yml file for env addons. Stay tuned!
